File: src/Graph.java
import java.util.*;

public class Graph {
    private final List<Edge> edges;
    private final Set<Vertice> vertices;
    private final Map<Vertice, Set<Vertice>> adjacencyMap;

    public Graph() {
        this.edges = new ArrayList<>();
        this.vertices = new HashSet<>();
        this.adjacencyMap = new HashMap<>();
    }

    public void addEdge(Edge edge) {
        edges.add(edge);
        updateVerticesAndConnections(edge);
    }

    public void removeEdge(Edge edge) {
        edges.remove(edge);
        updateVerticesAndConnectionsAfterRemoval(edge);
    }

    public Set<Vertice> getAdjacenciesCopy(Vertice vertice) {
        return new HashSet<>(adjacencyMap.getOrDefault(vertice, Collections.emptySet()));
    }

    public void addVertice(Vertice vertice) {
        vertices.add(vertice);
    }

    public void removeVertice(Vertice vertice) {
        vertices.remove(vertice);
        adjacencyMap.remove(vertice);

        for (Set<Vertice> connectedVertices : adjacencyMap.values()) {
            connectedVertices.remove(vertice);
        }
    }

    public Set<Vertice> getVerticesCopy() {
        return new HashSet<>(vertices);
    }

    public List<Edge> getEdgesCopy() {
        return new ArrayList<>(edges);
    }

    public Iterable<Vertice> verticesIterable() {
        return vertices::iterator;
    }

    public Iterable<Edge> edgesIterable() {
        return edges::iterator;
    }

    public Edge getEdge(int index) {
        return edges.get(index);
    }

    public int edgesSize() {
        return edges.size();
    }

    public boolean edgesIsEmpty() {
        return edges.isEmpty();
    }

    @Override
    public String toString() {
        return edges.toString();
    }

    private void updateVerticesAndConnections(Edge edge) {
        vertices.add(edge.vertex1());
        vertices.add(edge.vertex2());

        adjacencyMap
                .computeIfAbsent(edge.vertex1(), k -> new HashSet<>())
                .add(edge.vertex2());

        adjacencyMap
                .computeIfAbsent(edge.vertex2(), k -> new HashSet<>())
                .add(edge.vertex1());
    }

    private void updateVerticesAndConnectionsAfterRemoval(Edge edge) {
        vertices.remove(edge.vertex1());
        vertices.remove(edge.vertex2());

        Set<Vertice> connectedToV1 = adjacencyMap.get(edge.vertex1());
        Set<Vertice> connectedToV2 = adjacencyMap.get(edge.vertex2());

        if (connectedToV1 != null) {
            connectedToV1.remove(edge.vertex2());
            if (connectedToV1.isEmpty()) {
                adjacencyMap.remove(edge.vertex1());
            }
        }

        if (connectedToV2 != null) {
            connectedToV2.remove(edge.vertex1());
            if (connectedToV2.isEmpty()) {
                adjacencyMap.remove(edge.vertex2());
            }
        }
    }
}

File: versão javascript/Graph.js
import ObjectSet from './ObjectSet.js';
import ObjectMap from './ObjectMap.js';

export class Vertice {
    constructor(name) {
        this.name = name;
    }

    toString() {
        return `[v: ${this.name}]`;
    }

    toStringId() {
        return this.name;
    }
}

export class Edge {
    constructor(vertex1, vertex2) {
        this.vertex1 = vertex1;
        this.vertex2 = vertex2;
    }

    toString() {
        return `[v1: ${this.vertex1.name}, v2: ${this.vertex2.name}]`;
    }

    toStringId() {
        return `${this.vertex1.toStringId()}-${this.vertex2.toStringId()}`;
    }
}

export class Graph {
    constructor() {
        this.edges = new ObjectSet();
        this.vertices = new ObjectSet();
        this.adjacencyMap = new ObjectMap();
    }

    addEdge(edge) {
        this.edges.add(edge);
        this.updateVerticesAndConnections(edge);
    }

    removeEdge(edge) {
        this.edges.delete(edge);
        this.updateVerticesAndConnectionsAfterRemoval(edge);
    }

    getAdjacenciesCopy(vertice) {
        const connections = this.adjacencyMap.get(vertice);
        return connections ? connections.clone() : new ObjectSet();
    }

    addVertice(vertice) {
        this.vertices.add(vertice);
    }

    removeVertice(vertice) {
        this.vertices.delete(vertice);
        this.adjacencyMap.delete(vertice);

        for (const connectedVertices of this.adjacencyMap.values()) {
            connectedVertices.delete(vertice);
        }
    }

    getVerticesCopy() {
        return this.vertices.clone();
    }

    getEdgesCopy() {
        return this.edges.clone();
    }

    toString() {
        return [...this.edges].toString();
    }

    updateVerticesAndConnections(edge) {
        this.vertices.add(edge.vertex1);
        this.vertices.add(edge.vertex2);

        if (!this.adjacencyMap.has(edge.vertex1)) {
            this.adjacencyMap.set(edge.vertex1, new ObjectSet());
        }
        this.adjacencyMap.get(edge.vertex1).add(edge.vertex2);

        if (!this.adjacencyMap.has(edge.vertex2)) {
            this.adjacencyMap.set(edge.vertex2, new ObjectSet());
        }
        this.adjacencyMap.get(edge.vertex2).add(edge.vertex1);
    }

    updateVerticesAndConnectionsAfterRemoval(edge) {
        const removeIfEmpty = (map, key, value) => {
            const set = map.get(key);
            if (set) {
                set.delete(value);
                if (set.size() === 0) map.delete(key);
            }
        };

        removeIfEmpty(this.adjacencyMap, edge.vertex1, edge.vertex2);
        removeIfEmpty(this.adjacencyMap, edge.vertex2, edge.vertex1);
    }
}
